src/main/java/com/example/bookstore/service/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,95 @@
+import com.example.bookstore.model.Book;
+import com.example.bookstore.model.CartItem;
+import com.example.bookstore.model.ShoppingCart;
+import com.example.bookstore.model.User;
+import com.example.bookstore.repository.BookRepository;
+import com.example.bookstore.repository.CartItemRepository;
+import com.example.bookstore.repository.ShoppingCartRepository;
+import com.example.bookstore.repository.UserRepository;
+import com.example.bookstore.service.ShoppingCartService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final BookRepository bookRepository;
+    private final UserRepository userRepository;
+    private final CartItemRepository cartItemRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+
+    @Override
+    public ShoppingCartDto addBookToCart(AddCartItemRequestDto requestDto, Long userId) {
+        Long bookId = requestDto.getBookId();
+        Book book = bookRepository.findById(bookId)
+                .orElseThrow(
+                        () -> new EntityNotFoundException("Can't find book by id: " + bookId)
+                );
+
+        ShoppingCart shoppingCartFromDb = getOrCreateShoppingCart(userId);
+
+        CartItem cartItem = new CartItem();
+        cartItem.setQuantity(requestDto.getQuantity());
+        cartItem.setBook(book);
+        cartItem.setShoppingCart(shoppingCartFromDb);
+        cartItemRepository.save(cartItem);
+        shoppingCartFromDb.getCartItems().add(cartItem);
+        return shoppingCartMapper.toDto(shoppingCartFromDb);
+    }
+
+    @Override
+    public ShoppingCartDto getUserCart(Long userId) {
+        ShoppingCart shoppingCartFromDb = getOrCreateShoppingCart(userId);
+        return shoppingCartMapper.toDto(shoppingCartFromDb);
+    }
+
+    @Override
+    public ShoppingCartDto updateQuantity(
+            Long userId,
+            UpdateCartItemQuantityRequestDto requestDto,
+            Long id
+    ) {
+        CartItem cartItem = getCartItemById(id);
+        cartItem.setQuantity(requestDto.getQuantity());
+        cartItemRepository.save(cartItem);
+
+        ShoppingCart shoppingCartFromDb = getOrCreateShoppingCart(userId);
+        return shoppingCartMapper.toDto(shoppingCartFromDb);
+    }
+
+    @Override
+    public void deleteById(Long id) {
+        cartItemRepository.deleteById(id);
+    }
+
+    private ShoppingCart getOrCreateShoppingCart(Long userId) {
+        return shoppingCartRepository.findByUserId(userId)
+                .orElseGet(() -> {
+                    ShoppingCart shoppingCart = new ShoppingCart();
+                    shoppingCart.setUser(getUserById(userId));
+                    return shoppingCartRepository.save(shoppingCart);
+                });
+    }
+
+    private User getUserById(Long userId) {
```

Comment:
Shouldn't it be UserService responsibility? I would better use userService for it instead of adding this private method.